### PR TITLE
[rllib] Print out intermediate data shapes on the first iteration

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

To help new users understand the data flow in RLlib, this PR prints out important data types on the first iteration (adds less than 100 lines of output to a run of an agent). If the user does not want to see these values, they can set `"log_level": "WARN"`.

Here's an example of the additional output when initializing an agent:

```
(pid=26182) 2019-03-19 21:40:27,287	INFO policy_evaluator.py:689 -- Built policy map: {'default': <ray.rllib.agents.a3c.a3c_tf_policy_graph.A3CPolicyGraph object at 0x7f5c8a2980b8>}
(pid=26182) 2019-03-19 21:40:27,287	INFO policy_evaluator.py:690 -- Built preprocessor map: {'default': <ray.rllib.models.preprocessors.NoPreprocessor object at 0x7f5c8a283cf8>}
(pid=26183) 2019-03-19 21:40:34,850	INFO policy_evaluator.py:284 -- Creating policy evaluation worker 1 on CPU (please ignore any CUDA init errors)
(pid=26183) 2019-03-19 21:40:34.851496: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
(pid=26184) 2019-03-19 21:40:35,062	INFO policy_evaluator.py:284 -- Creating policy evaluation worker 2 on CPU (please ignore any CUDA init errors)
(pid=26184) 2019-03-19 21:40:35.063296: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
(pid=26183) 2019-03-19 21:40:35,378	INFO policy_evaluator.py:401 -- Generating sample batch of size 20
(pid=26183) 2019-03-19 21:40:35,378	INFO sampler.py:277 -- Initializing sampler
(pid=26183) 2019-03-19 21:40:35,379	INFO sampler.py:286 -- Unfiltered obs from env: { 0: { 'single_agent': np.ndarray((4,), dtype=float64, min=-0.011, max=0.049, mean=0.026)}}
(pid=26183) 2019-03-19 21:40:35,379	INFO sampler.py:287 -- Info return from env: {0: {'single_agent': None}}
(pid=26183) 2019-03-19 21:40:35,379	INFO sampler.py:372 -- Preprocessed obs: np.ndarray((4,), dtype=float64, min=-0.011, max=0.049, mean=0.026)
(pid=26183) 2019-03-19 21:40:35,379	INFO sampler.py:376 -- Filtered obs: np.ndarray((4,), dtype=float64, min=-0.011, max=0.049, mean=0.026)
(pid=26183) 2019-03-19 21:40:35,380	INFO sampler.py:484 -- Compute actions input:
(pid=26183) 
(pid=26183) { 'default': [ { 'data': { 'agent_id': 'single_agent',
(pid=26183)                            'env_id': 0,
(pid=26183)                            'info': None,
(pid=26183)                            'obs': np.ndarray((4,), dtype=float64, min=-0.011, max=0.049, mean=0.026),
(pid=26183)                            'prev_action': np.ndarray((), dtype=int64, min=0.0, max=0.0, mean=0.0),
(pid=26183)                            'prev_reward': 0.0,
(pid=26183)                            'rnn_state': []},
(pid=26183)                  'type': 'PolicyEvalData'}]}
(pid=26183) 
(pid=26183) 2019-03-19 21:40:35,398	INFO sampler.py:511 -- Compute actions result:
(pid=26183) 
(pid=26183) { 'default': ( np.ndarray((1,), dtype=int64, min=0.0, max=0.0, mean=0.0),
(pid=26183)                [],
(pid=26183)                { 'action_prob': np.ndarray((1,), dtype=float32, min=0.5, max=0.5, mean=0.5),
(pid=26183)                  'vf_preds': np.ndarray((1,), dtype=float32, min=-0.036, max=-0.036, mean=-0.036)})}
(pid=26183) 
(pid=26183) 2019-03-19 21:40:35,415	INFO sample_batch_builder.py:154 -- Postprocessed trajectory fragment:
(pid=26183) 
(pid=26183) { 'single_agent': { 'data': { 'action_prob': np.ndarray((16,), dtype=float32, min=0.498, max=0.502, mean=0.5),
(pid=26183)                               'actions': np.ndarray((16,), dtype=int64, min=0.0, max=1.0, mean=0.375),
(pid=26183)                               'advantages': np.ndarray((16,), dtype=float32, min=1.432, max=14.89, mean=8.335),
(pid=26183)                               'agent_index': np.ndarray((16,), dtype=int64, min=0.0, max=0.0, mean=0.0),
(pid=26183)                               'dones': np.ndarray((16,), dtype=bool, min=0.0, max=1.0, mean=0.062),
(pid=26183)                               'eps_id': np.ndarray((16,), dtype=int64, min=60683411.0, max=60683411.0, mean=60683411.0),
(pid=26183)                               'infos': np.ndarray((16,), dtype=object, head={}),
(pid=26183)                               'new_obs': np.ndarray((16, 4), dtype=float32, min=-0.814, max=1.74, mean=0.127),
(pid=26183)                               'obs': np.ndarray((16, 4), dtype=float32, min=-0.617, max=1.392, mean=0.112),
(pid=26183)                               'prev_actions': np.ndarray((16,), dtype=int64, min=0.0, max=1.0, mean=0.375),
(pid=26183)                               'prev_rewards': np.ndarray((16,), dtype=float32, min=0.0, max=1.0, mean=0.938),
(pid=26183)                               'rewards': np.ndarray((16,), dtype=float32, min=1.0, max=1.0, mean=1.0),
(pid=26183)                               't': np.ndarray((16,), dtype=int64, min=0.0, max=15.0, mean=7.5),
(pid=26183)                               'value_targets': np.ndarray((16,), dtype=float32, min=1.0, max=14.854, mean=8.089),
(pid=26183)                               'vf_preds': np.ndarray((16,), dtype=float32, min=-0.432, max=-0.036, mean=-0.246)},
(pid=26183)                     'type': 'SampleBatch'}}
(pid=26183) 
(pid=26183) 2019-03-19 21:40:35,436	INFO policy_evaluator.py:438 -- Completed sample batch:
(pid=26183) 
(pid=26183) { 'data': { 'action_prob': np.ndarray((20,), dtype=float32, min=0.498, max=0.502, mean=0.5),
(pid=26183)             'actions': np.ndarray((20,), dtype=int64, min=0.0, max=1.0, mean=0.4),
(pid=26183)             'advantages': np.ndarray((20,), dtype=float32, min=1.12, max=14.89, mean=7.164),
(pid=26183)             'agent_index': np.ndarray((20,), dtype=int64, min=0.0, max=0.0, mean=0.0),
(pid=26183)             'dones': np.ndarray((20,), dtype=bool, min=0.0, max=1.0, mean=0.05),
(pid=26183)             'eps_id': np.ndarray((20,), dtype=int64, min=60683411.0, max=143039766.0, mean=77154682.0),
(pid=26183)             'infos': np.ndarray((20,), dtype=object, head={}),
(pid=26183)             'new_obs': np.ndarray((20, 4), dtype=float32, min=-0.814, max=1.74, mean=0.103),
(pid=26183)             'obs': np.ndarray((20, 4), dtype=float32, min=-0.617, max=1.392, mean=0.091),
(pid=26183)             'prev_actions': np.ndarray((20,), dtype=int64, min=0.0, max=1.0, mean=0.35),
(pid=26183)             'prev_rewards': np.ndarray((20,), dtype=float32, min=0.0, max=1.0, mean=0.9),
(pid=26183)             'rewards': np.ndarray((20,), dtype=float32, min=1.0, max=1.0, mean=1.0),
(pid=26183)             't': np.ndarray((20,), dtype=int64, min=0.0, max=15.0, mean=6.3),
(pid=26183)             'value_targets': np.ndarray((20,), dtype=float32, min=1.0, max=14.854, mean=6.97),
(pid=26183)             'vf_preds': np.ndarray((20,), dtype=float32, min=-0.432, max=0.13, mean=-0.194)},
(pid=26183)   'type': 'SampleBatch'}
(pid=26183) 
(pid=26182) 2019-03-19 21:40:35,643	INFO policy_evaluator.py:529 -- Training on batch:
(pid=26182) 
(pid=26182) { 'data': { 'action_prob': np.ndarray((40,), dtype=float32, min=0.498, max=0.502, mean=0.5),
(pid=26182)             'actions': np.ndarray((40,), dtype=int64, min=0.0, max=1.0, mean=0.4),
(pid=26182)             'advantages': np.ndarray((40,), dtype=float32, min=1.12, max=18.176, mean=8.614),
(pid=26182)             'agent_index': np.ndarray((40,), dtype=int64, min=0.0, max=0.0, mean=0.0),
(pid=26182)             'dones': np.ndarray((40,), dtype=bool, min=0.0, max=1.0, mean=0.05),
(pid=26182)             'eps_id': np.ndarray((40,), dtype=int64, min=60683411.0, max=262885849.0, mean=170020265.5),
(pid=26182)             'infos': np.ndarray((40,), dtype=object, head={}),
(pid=26182)             'new_obs': np.ndarray((40, 4), dtype=float32, min=-1.141, max=1.744, mean=0.078),
(pid=26182)             'obs': np.ndarray((40, 4), dtype=float32, min=-1.141, max=1.744, mean=0.068),
(pid=26182)             'prev_actions': np.ndarray((40,), dtype=int64, min=0.0, max=1.0, mean=0.35),
(pid=26182)             'prev_rewards': np.ndarray((40,), dtype=float32, min=0.0, max=1.0, mean=0.925),
(pid=26182)             'rewards': np.ndarray((40,), dtype=float32, min=1.0, max=1.0, mean=1.0),
(pid=26182)             't': np.ndarray((40,), dtype=int64, min=0.0, max=19.0, mean=7.9),
(pid=26182)             'value_targets': np.ndarray((40,), dtype=float32, min=1.0, max=18.209, mean=8.417),
(pid=26182)             'vf_preds': np.ndarray((40,), dtype=float32, min=-0.432, max=0.158, mean=-0.197)},
(pid=26182)   'type': 'SampleBatch'}
(pid=26182) 
(pid=26182) 2019-03-19 21:40:35,750	INFO policy_evaluator.py:551 -- Training output:
(pid=26182) 
(pid=26182) { 'stats': { 'cur_lr': 9.999999747378752e-05,
(pid=26182)              'grad_gnorm': 40.0,
(pid=26182)              'model': {},
(pid=26182)              'policy_entropy': 27.725748,
(pid=26182)              'policy_loss': 238.86201,
(pid=26182)              'var_gnorm': 22.64951,
(pid=26182)              'vf_explained_var': 0.029580772,
(pid=26182)              'vf_loss': 1980.9261}}
(pid=26182) 
```
## Related issue number

https://github.com/ray-project/ray/issues/4418

### CODE REVIEW

**Comment by ericl** [comment]
ptal

File: python/ray/rllib/utils/debug.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    in RLlib, such as SampleBatch, numpy arrays, etc.
+    """
+
+    return _printer.pformat(_summarize(obj))
+
+
+def _summarize(obj):
+    if isinstance(obj, dict):
+        return {k: _summarize(v) for k, v in obj.items()}
+    elif hasattr(obj, "_asdict"):
```
Comment (hartikainen) [inline python/ray/rllib/utils/debug.py]: What objects have this?

**Review by hartikainen — APPROVED** [review]
lgtm

File: python/ray/rllib/utils/debug.py (hunk 1 above)
Comment (ericl) [reply python/ray/rllib/utils/debug.py]: These would be namedtuples

Comment (hartikainen) [reply python/ray/rllib/utils/debug.py]: Would it make sense to check `isinstance(obj, namedtuple)`? Might make it a bit clearer.

**Pull request merged**